### PR TITLE
feat(core)!: two-phase broadcast status + reconcile grace window

### LINKED ISSUE #25: Harden sponsor ledger state transitions: two-phase broadcast status + reconcile grace window
## Context

`@aibtc/tx-schemas@0.8.0` shipped the runtime helpers (`classifyOccupant`, `decideBroadcast`, `adoptOrphan`, `quarantine`, `reconcile`, `SponsorLedgerSchema`) that `x402-sponsor-relay#338` is waiting on. Before the relay integration lands on **mainnet** traffic, two residual state-confidence gaps need to close in the schema layer so downstream consumers don't each reinvent the fix.

Both are about closing windows where sponsor-wallet state becomes briefly ambiguous. Neither is theoretical — they are the last two reasons the relay refactor can't claim full confidence in sponsor-wallet state at all times.

## Gap 1: single-phase broadcast status

Today `LedgerEntry.status` transitions straight to `broadcast_sent` at the moment of the ledger write. That implicitly assumes the write and the `broadcastTransaction` network call complete atomically.

On Cloudflare Workers / Durable Objects, edge termination between the storage write and the network-call return is a real window. The ledger can hold an entry claiming `broadcast_sent` for a tx that never hit the node — or conversely, a crash before the status update can leave a row without a clear outcome. Either way the next alarm cycle sees an entry reconcile cannot confidently classify.

### Proposal

Add a `pending_broadcast` state to `SponsorLedgerSchema`. Canonical transition:

```
pending_broadcast → broadcast_sent    (network call returned success)
pending_broadcast → broadcast_failed  (network call returned error)
pending_broadcast → <timeout>         (no resolution within N; reconcile handles)
```

Consumers write `pending_broadcast` **before** the network call, resolve on return, and reconcile sweeps stragglers. Document the two-phase write pattern in the README.

## Gap 2: no grace window in `reconcile()`

`reconcile()` compares the ledger against a mempool snapshot to classify entries. A tx broadcast seconds ago may not yet appear in Hiro's mempool index — but it is already in flight (the node accepted it). Classifying it as missing in that window risks premature adoption, quarantine, or re-broadcast.

Nakamoto blocks are 3-5s, so the real ambiguity is small (6-10 blocks + indexer lag, ~30s max), but it is not zero and cannot be defined away at the consumer.

### Proposal

Add a configurable grace window to `reconcile()`:

```ts
reconcile(ledger, mempoolReadByNonce, { justBroadcastGraceSeconds: 30 })
```

- Ledger entries with `broadcast_at` within the grace window and no mempool occupant: classified as `in_flight_pending_index` (or similar), **not** `missing`.
- Default grace: 30s. Override per consumer.
- Document the semantics in the README so consumers understand when they still need their own backoff vs. when the schema handles it.

## Tasks

- Add `pending_broadcast` to `SponsorLedgerSchema.status` enum; define transition rules.
- Add `broadcast_at: number` (unix ms) to `LedgerEntry` if not already present (reconcile needs it for grace-window logic).
- Add `justBroadcastGraceSeconds` option to `reconcile()`; default 30.
- Add `in_flight_pending_index` (or chosen name) variant to `ReconcileResult` classification union.
- Update `classifyOccupant` / `decideBroadcast` to handle the new state cleanly (a `pending_broadcast` entry should not be treated as a confirmed sponsor-owned slot for RBF decisions).
- Workflow tests:
  - Two-phase write happy path: `pending_broadcast` → `broadcast_sent`
  - Two-phase write failure: `pending_broadcast` → `broadcast_failed`
  - Two-phase write timeout: entry still in `pending_broadcast` past reasonable bound → reconcile detects and resolves
  - Grace window: ledger entry broadcast 10s ago, not in mempool → classified `in_flight_pending_index`, not `missing`
  - Grace window expiry: ledger entry broadcast 60s ago, not in mempool → classified `missing` (or the appropriate terminal variant)
- README section: "Canonical write pattern: two-phase broadcast"

## Acceptance criteria

- `SponsorLedgerSchema` includes `pending_broadcast` and `broadcast_at`.
- `reconcile()` accepts a grace-window option; tests cover inside-window and outside-window classification.
- Published minor (`0.9.0`) consumable by the relay.
- README documents the two-phase write contract with a minimal example.

## Out of scope

- Relay integration — tracked in [aibtcdev/x402-sponsor-relay#338](https://github.com/aibtcdev/x402-sponsor-relay/issues/338) and the `sponsor-ledger-integration` quest.
- Reorg handling — genuine irreducible concern; not addressed by this schema change.

## Why this lands before the relay refactor completes

The relay work is gated on mainnet-grade state confidence. Without these two closures, the relay would either:
1. Reinvent the same fix inline (defeats the point of the tx-schemas integration), or
2. Ship the refactor knowing state is ambiguous in ~0.1% of broadcasts — unacceptable for mainnet funds.

Small schema change; large downstream confidence gain.

### PR DESCRIPTION
**Pull request by whoabuddy**
Closes #25.

## Summary

- Introduces `LedgerEntry.status` (`pending_broadcast` | `broadcast_sent` | `broadcast_failed`) with explicit lifecycle helpers — `beginPendingBroadcast` / `resolveBroadcast` — backed by a transition matrix enforced at the schema layer (`LedgerTransitionError` on invalid moves).
- Adds `justBroadcastGraceSeconds` (default 30) to `reconcile()`; ledger entries absent from the mempool but broadcast within the grace window are surfaced as `inFlightPendingIndex` instead of `dropped`. `reconcile()` also auto-promotes `pending_broadcast` → `broadcast_sent` on mempool confirmation.
- `decideBroadcast` gains an `await_pending_broadcast` variant that short-circuits any other decision while a ledger entry is unresolved — hard invariant against silent double-broadcast.

## Design notes

- **`OccupantClassification` stays closed.** Grace-window ambiguity needs ledger + clock context that `classifyOccupant` has neither of by design; forcing it in would make every caller grace-aware for a concern only `reconcile` can answer. `inFlightPendingIndex` is additive on `ReconcileResult` where that context lives.
- **Timestamps stay ISO.** `broadcastAt` remains `IsoDateTimeSchema`; grace math parses to epoch ms internally. No parallel unix-ms field.
- **Lifecycle helpers over README conventions.** A two-phase write "pattern" documented only in prose drifts. `beginPendingBroadcast` / `resolveBroadcast` enforce the transition matrix at the type boundary and throw on misuse.

## Breaking change

`SponsorLedgerEntry.status` is required. Entries persisted under `0.8.0` lack it and will fail schema parse. Consumers must backfill `status: "broadcast_sent"` on migration. Targets `0.9.0`.

## Test plan

- [x] 217 tests pass (`npm test`)
- [x] `npm run typecheck` clean
- [x] `npm run build` clean
- [x] New coverage: every state transition (valid + invalid), `await_pending_broadcast` precedence over RBF, grace window inside/outside (10s/60s) and custom override (45s × {30, 60}), `reconcile` pending-promotion, crash-recovery scenario
- [ ] Relay integration smoke (out of scope here — tracked in aibtcdev/x402-sponsor-relay#338)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/core/sponsor-wallet-machine.ts [hunk 1]
```diff
@@ -7,8 +7,13 @@ import {
   StacksAddressSchema,
   TransactionIdSchema,
 } from "./primitives.js";
-import { getLedgerEntry, SponsorLedgerEntrySchema } from "./sponsor-ledger.js";
+import {
+  getLedgerEntry,
+  LedgerEntryStatusSchema,
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply src/core/sponsor-wallet-machine.ts]: Fixed in 3b1ec6f — removed the unused import.

File: README.md [hunk 2]
```diff
@@ -153,3 +153,68 @@ const { wallet: next, ledger: nextLedger, adopted, dropped } = reconcile(
...
+  ledger = resolveBroadcast(ledger, nonce, "sent", { lastOutcome: outcome });
+} catch (err) {
+  ledger = resolveBroadcast(ledger, nonce, "failed");
+  throw err;
+}
+```
+
+`decideBroadcast` refuses to issue a new decision while the entry is
+`pending_broadcast` — it returns `{ kind: "await_pending_broadcast", nonce,
+txId }` so the consumer resolves the prior call before a second broadcast
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply README.md]: Fixed in 3b1ec6f — reformatted as a fenced block so renderers don't split the literal across a newline.

File: src/core/sponsor-wallet-machine.ts [hunk 3]
```diff
@@ -443,11 +616,19 @@ export function reconcile(
...
     const observed = mempoolReadByNonce[nonce];
-    if (!observed || observed.tx_id !== entry.txId) {
+    if (observed && observed.tx_id === entry.txId) continue;
+
+    const ageMs = nowMs - Date.parse(entry.broadcastAt);
+    if (ageMs >= 0 && ageMs < graceMs) {
+      inFlightPendingIndex.push(entry.nonce);
+    } else {
       dropped.push(entry.nonce);
     }
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply src/core/sponsor-wallet-machine.ts]: Fixed in 3b1ec6f — sharp catch. Drift and propagation lag are different signals; they shouldn't share the grace window. `reconcile` now: grace-window applies only when `observed` is null/undefined; any `observed.tx_id !== entry.txId` is classified as `dropped` regardless of age. Added a test (`treats txId drift as dropped even within the grace window`) that broadcasts 5s ago and still ends up in `dropped` when the mempool reports a different sponsor-owned tx at that nonce.

**Review by arc0btc — APPROVED** [review]
Two-phase broadcast lifecycle + reconcile grace window — closes the crash-window gap that's been an open reliability concern for the relay.

**What works well:**
- The transition matrix is airtight: `PENDING_PREDECESSORS` set clearly models which states are valid predecessors, and `LedgerTransitionError` gives callers a typed error to catch rather than a silent wrong-state write.
- `decideBroadcast` guarding at the top — returning `await_pending_broadcast` before any other logic — is the right precedence call. It keeps the invariant clean regardless of occupant classification.
- The two-pass structure in `reconcile()` is well-commented and the separation of concerns is sound: Pass 1 handles mempool hits (adopt + promote), Pass 2 handles absences (grace vs. dropped). The second commit's fix making txId-drift unconditionally `dropped` is exactly right — drift is not lag.
- Test coverage is thorough. The crash-recovery scenario (entry broadcast 120s ago, never resolved, not in mempool → dropped) and the grace-window boundary tests (10s inside / 60s outside / custom override) give me confidence the edge cases are understood.

**[question] Grace window applies to `broadcast_sent` entries too** (`src/core/sponsor-wallet-machine.ts`, Pass 2)

Pass 2 computes `ageMs` from `entry.broadcastAt` for all entries when `observed` is null — regardless of whether the entry is `pending_broadcast` or `broadcast_sent`. A `broadcast_sent` entry that temporarily disappears from the Hiro mempool (say, due to rapid confirmation or transient indexer lag) within 30s of its `broadcastAt` timestamp would land in `inFlightPendingIndex` rather than `dropped`. Is this intentional? Operationally, a `broadcast_sent` entry going absent is a different signal than a `pending_broadcast` one — the former means the node confirmed receipt, the latter means we haven't heard back yet. If the intent is grace-for-pending-only, the condition could be:

```suggestion
    const ageMs = nowMs - Date.parse(entry.broadcastAt);
    if (entry.status === "pending_broadcast" && ageMs >= 0 && ageMs < graceMs) {
      inFlightPendingIndex.push(entry.nonce);
    } else {
```

If the current behavior is intentional (grace for any recent absence), a brief comment explaining why would help future readers.

**[question] `justBroadcastGraceSeconds` default vs. stated propagation timing** (`src/core/sponsor-wallet-machine.ts:507`)

The comment says the grace window "Covers node→indexer propagation lag (~6-10 Nakamoto blocks + Hiro indexing)". If those are streaming Stacks blocks at ~1-2s each, 30s is comfortably generous. If they're longer, 30s could be tight. The x402 relay can override this via `ReconcileOptions`, but for consumers who don't override, it's worth confirming the 30s default is calibrated to real Hiro indexing latency. We've seen Hiro 400 delays operationally, though those are different failure modes.

**[nit] Variable shadowing in README example** (`README.md`)

```ts
let ledger = beginPendingBroadcast(ledger, {
```

This shadows the outer `ledger` binding. JS evaluates the RHS correctly so the code works, but the pattern is confusing to read — looks like a circular assignment at a glance. `let pendingLedger = beginPendingBroadcast(ledger, ...)` or a reassignment (`ledger = beginPendingBroadcast(ledger, ...)`) would read more naturally.

**Code quality notes:**

The `rbfAttempts` defaulting chain in `beginPendingBroadcast` (`input.rbfAttempts ?? existing?.rbfAttempts ?? 0`) is correct but callers doing RBF need to remember to pass the incremented count explicitly — there's no auto-increment. The README covers the first-broadcast case but not the RBF flow. A brief README addition showing the RBF call (`rbfAttempts: existingEntry.rbfAttempts + 1`) would prevent a subtle foot-gun for consumers integrating this pattern.

**Operational context:**

We run this code path nightly for Zest sBTC supply operations. The nonce-serialization fix we shipped last week (shared nonce coordinator) prevents concurrent double-submissions at the application layer, but there was still a theoretical crash window between write and node confirmation. This closes it cleanly at the state machine layer. The `await_pending_broadcast` guard is the piece we'd want most — if the process dies mid-broadcast, crash recovery via reconcile picks it up without a second broadcast firing.

The two questions above are worth a quick check before merge, but neither is blocking. The core logic is sound and the test coverage gives me confidence in the critical paths.

**Comment by whoabuddy** [comment]
Thanks for the thorough read, @arc0btc — all four points addressed in a909bbc.

**[question 1] Grace window leaking to broadcast_sent — fixed.**
Sharp catch. You're right that `broadcast_sent` and `pending_broadcast` carry different operational meaning when they vanish from the mempool. Pass 2 now branches explicitly:

- `pending_broadcast` + absent → grace window applies (could still be propagation lag)
- `broadcast_sent` + absent → `dropped` immediately (node confirmed receipt; absence means mined or evicted, caller disambiguates via tx status lookup)
- any `tx_id` mismatch → `dropped` immediately (drift, not lag)

Added a test `does NOT grace-window broadcast_sent entries` that pins this: 5s-old `broadcast_sent` entry, mempool null → `dropped`, not `inFlightPendingIndex`. Comment on `ReconcileOptions.justBroadcastGraceSeconds` now spells out the scope.

**[question 2] 30s default — keeping it, but documented honestly.**
Stayed at 30s to match the issue spec, but expanded the comment to acknowledge it's calibrated to the tight end of the Nakamoto-block + Hiro-indexer envelope (3-5s × 6-10 blocks ≈ 30-50s). Operators seeing routine Hiro slowness should override upward; the grace only delays a `dropped` classification, never changes a true outcome. If we hit operational pain on this default, we can revisit in a 0.10.x.

**[nit 3] README shadowing — fixed.**
Dropped the `let ledger = ...` rebind in favor of straight reassignment. Reads as the iterative state update it actually is.

**[code quality] RBF flow example — added.**
Good foot-gun call. README now has a second snippet showing the explicit `rbfAttempts: existing.rbfAttempts + 1` pass on retry, with a note that `beginPendingBroadcast` doesn't auto-increment because the same helper covers first broadcasts and retries.

Operationally: the Zest sBTC nightly path is exactly the consumer this is meant to harden. The shared nonce coordinator catches concurrent submissions; this catches the in-process crash window. Together that's the gap closed end-to-end. 219 tests green, build clean — ready for merge when you are.
